### ISSUE TITLE
Blocked Host with system test

### ISSUE THREAD

**Issue by adrienpoly** [rails/rails#37474]
### Steps to reproduce
I have a docker-compose environment similar to 
https://github.com/bibendi/dip-example-rails
To this environment, I have added a `selenium_chrome` service 

```yml
selenium_chrome:
    image: selenium/standalone-chrome-debug
    logging:
      driver: none
    ports:
      - '5900:5900'
```
And system tests are ran using Rspec/Capybara

```ruby
# spec/support/capybara.rb
require 'selenium-webdriver'
require 'capybara/rails'

Capybara.server = :puma, { Silent: true }

Capybara.register_driver :selenium_chrome_in_container do |app|
  Capybara::Selenium::Driver.new(
    app,
    browser: :remote,
    url: 'http://selenium_chrome:4444/wd/hub',
    desired_capabilities: :chrome
  )
end

RSpec.configure do |config|
  config.before(:each, type: :system) do
    driven_by :selenium_chrome_in_container
    Capybara.server_host = '0.0.0.0'
    Capybara.server_port = '3000'
    Capybara.app_host = 'http://rails:3000'
  end
end
```

### Expected behavior
I would like to be able to run system tests in a docker-compose environment

### Actual behavior

My remote browser can't access the rails server

<img width="1050" alt="Screen Shot 2019-10-15 at 11 10 29 AM" src="https://user-images.githubusercontent.com/7847244/66818218-4ae3d700-ef3d-11e9-88e3-8bd770bb6174.png">

in test.rb I have tested the following lines with always the same result

- `config.hosts << 'rails'`
- `config.hosts = nil`
- `config.hosts.clear`

Whatever value I pass to config host I always get the same error

Strangely if I do a byebug in my test and get the value of `config.hosts` it is correctly set

I am probably missing something here but after spending hours on it I thought I would raise this issue

Thanks

### System configuration
Rails 6.0.0
Ruby 2.6.5
Rspec 4.0.0.beta2

**Comment by eileencodes** [comment]
Can you reproduce this with minitest instead of Rspec?

**Comment by JuanitoFatas** [comment]
A workaround before the root cause revealed would be adding the following to `config/environments/test.rb`

```ruby
config.hosts << "www.example.com"
``` 

Because [Capybara tries to access `www.example.com`] (either in request / feature specs).

[Capybara tries to access `www.example.com`]: https://github.com/teamcapybara/capybara#gotchas

**Comment by gsamokovarov** [comment]
What's the environment in which this test is run at? Is `config.hosts` already set when the tests run?

**Comment by Auralcat** [comment]
@adrienpoly Are you still experiencing this problem with newer Rails versions or with edge Rails? Could you reproduce this issue with Minitest instead of RSpec?

**Comment by viktorianer** [comment]
> @adrienpoly Are you still experiencing this problem with newer Rails versions or with edge Rails? Could you reproduce this issue with Minitest instead of RSpec?

I do have 2 different projects for my students. One with Minitest and one with RSpec. Those projects nearly identical. Only running RSpec inside Docker produce this error:

The error message from rspec:

```
UsersSignUpInterface signup form renders all input fields
      Failure/Error: assert_template "users/new"
      
      Minitest::Assertion:
        expecting <"users/new"> but rendering with <["rescues/blocked_host", "rescues/layout"]>
      # /usr/local/bundle/gems/minitest-5.13.0/lib/minitest/assertions.rb:183:in `assert'
      # /usr/local/bundle/gems/rails-controller-testing-1.0.4/lib/rails/controller/testing/template_assertions.rb:131:in `assert_template'
      # ./spec/requests/users_sign_up_interface_spec.rb:86:in `block (3 levels) in <main>'

```

Running Minitest in- or outside Docker and RSpec outside Docker produce no error. 

**System configuration**
ruby 2.6.5
rails (6.0.2.1)
rspec (3.9.0)
Docker version 19.03.8, build afacb8b
docker-compose version 1.25.4, build 8d51620a

**Comment by rainkinz** [comment]
I had the same issue, and realized the issue was that `RAILS_ENV` was being set to `development` in my `docker-compose.yml` for the container by default:

```
    environment:
      - NODE_ENV=development
      - RAILS_ENV=${RAILS_ENV:-development}
      - REDIS_URL=redis://redis:6379/
```

This was causing the tests to run using the `development` environment! Oops 🤦  . Anyway, removing `- RAILS_ENV=${RAILS_ENV:-development}` fixed the issue. Not sure if it's the same for you, but thought I'd mention it.

**Comment by gsamokovarov** [comment]
I believe the issue is somewhere around the env. The hosts blocking is enabled only in `development` by default or without any explicit configuration.

**Comment by jonathanhefner** [comment]
> I believe the issue is somewhere around the env.

That is my guess too.  The default generated `test_helper.rb` sets [`ENV['RAILS_ENV'] ||= 'test'`](https://github.com/rails/rails/blob/v6.0.2.2/railties/lib/rails/generators/rails/app/templates/test/test_helper.rb.tt#L1).  If `ENV['RAILS_ENV']` is already set, or if `test_helper.rb` is not loaded (does RSpec have an equivalent?), then you will probably encounter this issue.

As an alternative to changing `docker-compose.yml`, you might consider changing `ENV['RAILS_ENV'] ||= 'test'` to simply `ENV['RAILS_ENV'] = 'test'`.  Whatever works best for your setup.

Also, it's worth noting that apps (and plugins) generated by Rails 5.1 did not include this line in `test_helper.rb`.  It was removed for 5.1, but then restored for 5.2 by #31355.

**Comment by viktorianer** [comment]
Hm, not sure just about the env. I run my tests/specs in a test container with:

```
version: '3.7'

services:
  app: &app
    image: have_a_car:0.0.1
    env_file: .env
  ...
...
  test:
    <<: *app
    command: rspec
    environment:
      RAILS_ENV: test
      NODE_ENV: test
  ...
...
``` 

And the `.env` file contains:

```
...
RAILS_ENV=development
NODE_ENV=development
...
```

So, it is possible. 

But, I installed some updates over the week and restarted my mac. And I just opened the project again, but everything works fine! -_- I did not change any line of code.

Strange, but system configuration still the same! However, I did remove all games ones (`gem uninstall all`) and reinstalled Bundler (`gem install bundler`) and all gems in project (`bundle install`).

**Comment by jlefley** [comment]
I have a similar setup using Rails, Rspec, and Docker with a selenium_chrome service for testing. The system test were working fine using default configurations and not setting config.hosts.

Similar to the original reporter of this issue, I started getting the blocked host error when running system tests through the selenium_chrome Docker service out of what seemed like nowhere.

I checked my env by printing it directly from the spec file and I am definitely running rails in test mode.

I also printed out Rails.application.config.hosts in the spec file and it was an empty array.

Adding the Docker container host name to config.hosts in config/environments/test.rb did not make any difference. I had to add the Docker container host name to config.hosts in config/environments/development.rb to correct the error and get the tests working again.

This makes no sense why a setting in the development environment would be applying to the test environment. Again, I printed the rails env and config.hosts in my spec file after adding the host to the development config and it still says test and an empty array.

It's working now but I am confused why a development mode setting is affecting the test environment like this.

Rails 6.0.2.1

**Comment by NicoShanstrom** [comment]
in rails_helper.rb i took the advice from above to do this: ENV['RAILS_ENV'] = 'test'
# ENV['RAILS_ENV'] ||= 'test' from @jonathanhefner and all my tests are passing again. I just wonder if i will have to modify something else again if in development or production?

**Comment by clickworkorange** [comment]
All my controller tests (minitest) broke after I set `config.hostname` in `environments/test.rb`, giving me a `403: Forbidden` on every request. Looking at the test log revealed a bit more information: 
```
[ActionDispatch::HostAuthorization::DefaultResponseApp] Blocked hosts: www.example.com
```
At first I had no idea where this `www.example.com` domain came from, but I eventually found that a `DEFAULT_HOST` constant [is defined](https://github.com/rails/rails/blob/3235827585d87661942c91bc81f64f56d710f0b2/actionpack/lib/action_dispatch/testing/integration.rb#L92) in `actionpack-8.0.2/lib/action_dispatch/testing/integration.rb` with the value `"www.example.com"`, which I assume is the root cause. This was surprising behaviour for two reasons: firstly, because I would expect the values set in the environment file to be respected, secondly because I wasn't able to find any information about how or why this constant is used in integration tests. 

For now I used @JuanitoFatas [work-around](https://github.com/rails/rails/issues/37474#issuecomment-559974661), adding "www.example.com" to `config.hosts`, which solved the issue, but it would be nice to understand what's actually going on!
